### ISSUE TITLE
Performance: Model pruning and quantization-aware training for YOLOX

### ISSUE THREAD

**Issue by filthyrake** [filthyrake/telescope_cam_detection#165]
## Overview
Apply advanced model optimization: structured pruning and quantization-aware training (QAT) for better INT8 performance.

## Motivation
Post-training quantization (PTQ) with TensorRT INT8 gives ~2x speedup, but QAT can improve:
- Better INT8 accuracy (less degradation)
- Structured pruning: 25% parameter reduction, 18% FPS increase, 0.5% mAP loss (proven)
- Smaller model size for faster loading/inference

### 1. Structured Pruning
Remove entire filters/channels (not just weights):
```python
import torch_pruning as tp

# Define pruning strategy
strategy = tp.strategy.L1Strategy()  # Prune by L1 norm

# Prune model
pruner = tp.pruner.MetaPruner(
    model,
    example_inputs,
    importance=strategy,
    pruning_ratio=0.25,  # Remove 25% of parameters
)
pruned_model = pruner.prune()

# Fine-tune on wildlife dataset
fine_tune(pruned_model, wildlife_dataset, epochs=10)
```

**Expected**: 25% fewer parameters, 18% FPS increase, <1% mAP loss

### 2. Quantization-Aware Training (QAT)
Train model with quantization ops in the loop:
```python
import torch.quantization

# Prepare model for QAT
model.qconfig = torch.quantization.get_default_qat_qconfig('fbgemm')
model_prepared = torch.quantization.prepare_qat(model)

# Train with quantization (fake quantization during training)
train(model_prepared, wildlife_dataset, epochs=5)

# Convert to fully quantized INT8
model_int8 = torch.quantization.convert(model_prepared)
```

**Expected**: INT8 with <1% mAP loss (vs. 2-3% with PTQ)

### 3. Knowledge Distillation
Train smaller "student" model to mimic large "teacher":
```python
# Teacher: YOLOX-x (large, accurate)
teacher = YOLOXLarge()

# Student: YOLOX-tiny (small, fast)
student = YOLOXTiny()

# Distillation loss
loss = distillation_loss(
    student_output, 
    teacher_output,
    temperature=3.0
)
```

**Expected**: Faster model with teacher's accuracy

### Phase 1: Dataset Preparation
- Collect 5,000 wildlife frames from cameras
- Annotate with existing detections (pseudo-labels)
- Split: 80% train, 10% val, 10% test

### Phase 2: Pruning
1. Prune YOLOX by 25% (structured pruning)
2. Fine-tune on wildlife dataset (10 epochs)
3. Validate accuracy on test set
4. Convert to TensorRT and benchmark

### Phase 3: QAT
1. Start from pruned model
2. QAT training (5 epochs)
3. Convert to INT8
4. Compare to PTQ INT8

### Phase 4: Evaluation
- Benchmark speed vs. accuracy trade-offs
- Test on live camera streams
- A/B test with users

## Expected Results

| Model | mAP | FPS | Model Size | VRAM |
|-------|-----|-----|------------|------|
| YOLOX-s (baseline) | 40.5% | 30 | 35MB | 800MB |
| YOLOX-s + Pruning | 40.0% | 35 | 26MB | 600MB |
| YOLOX-s + QAT INT8 | 40.0% | 60 | 9MB | 400MB |
| YOLOX-tiny (distilled) | 38.0% | 80 | 5MB | 300MB |

## Configuration
```yaml
detection:
  model:
    optimization:
      pruning:
        enabled: true
        ratio: 0.25
        strategy: "l1"  # or "l2", "taylor"
        
      qat:
        enabled: true
        epochs: 5
        backend: "fbgemm"  # or "qnnpack" for mobile
        
      distillation:
        teacher_model: "yolox-x"
        student_model: "yolox-tiny"
        temperature: 3.0
```

## Implementation Tasks
- [ ] Collect and annotate wildlife training dataset
- [ ] Implement structured pruning pipeline
- [ ] Fine-tune pruned model
- [ ] Implement QAT training
- [ ] Compare pruned vs. QAT vs. baseline
- [ ] Document optimal configuration

## Validation
1. Test on 1000 wildlife images (annotated)
2. Compare mAP: pruned vs. baseline
3. Benchmark FPS: PyTorch and TensorRT
4. Live camera test: 24-hour deployment

## Success Metrics
- [ ] Pruned model: ≥15% FPS gain, <1% mAP loss
- [ ] QAT INT8: ≥2x FPS gain, <1% mAP loss (better than PTQ)
- [ ] Smaller model size (faster loading, less VRAM)

## Priority
**LOW** - Advanced optimization, requires custom training pipeline

## Related Issues
- Builds on TensorRT INT8 conversion (#157)
- Requires wildlife training dataset collection

## References
- Structured pruning paper: "Model Optimization Techniques for YOLO Models"
- QAT tutorial: https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html
- Distillation: https://arxiv.org/abs/1503.02531
- torch-pruning: https://github.com/VainF/Torch-Pruning

## Timeline
- Dataset collection: 1 week
- Pruning experiments: 1 week
- QAT training: 1 week
- Evaluation: 1 week
- **Total: 1 month**
